### PR TITLE
migrator: initialize migrator

### PR DESCRIPTION
**Pull request by drjova**
* INSPIR-2208

Signed-off-by: Harris Tzovanakis <me@drjova.com>

### CODE REVIEW

File: inspirehep/accounts/api.py [hunk 1]
```diff
@@ -26,3 +26,14 @@ def is_superuser_or_cataloger_logged_in():
         user_roles = {role.name for role in current_user.roles}
         return user_roles & {"superuser", "cataloger"}
     return False
+
+
+def login_required_with_role_superadmin_or_cataloger(func):
```
Comment (harunurhan) [inline inspirehep/accounts/api.py]: maybe it's better to have sth `login_required_with_roles` or have `roles` param for the `login_required` that defaults to `all`/`user`

File: inspirehep/config.py [hunk 2]
```diff
@@ -152,6 +145,7 @@ def _(x):
 LEGACY_RECORD_URL_PATTERN = "https://inspirehep.net/record/{recid}"
 AUTHENTICATION_TOKEN = "CHANGE_ME"
 INSPIRE_NEXT_URL = "http://web-next:5000"
+LEGACY_BASE_URL = "http://inspirehep.net"
```
Comment (harunurhan) [inline inspirehep/config.py]: `https` ?

File: inspirehep/migrator/dumper.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+# the terms of the MIT License; see LICENSE file for more details.
+
+"""Migrator dumper."""
+
+from marshmallow import ValidationError
+
+from .serializers.schemas.json import ErrorList
+
+
+def marshmallow_dumper(schema_class):
```
Comment (harunurhan) [inline inspirehep/migrator/dumper.py]: we should perhaps get rid of this.

File: pyproject.toml [hunk 4]
```diff
@@ -62,6 +65,7 @@ pylint = "*"
 flake8 = "*"
 watchdog = "*"
 virtualenv = "<16.5.0"
+ipdb = "^0.12.0"
```
Comment (harunurhan) [inline pyproject.toml]: thanks

File: inspirehep/accounts/api.py (hunk 1 above)
Comment (drjova) [reply inspirehep/accounts/api.py]: done

**Comment by drjova** [comment]
@harunurhan updated

File: inspirehep/migrator/tasks.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        ensure_valid_schema(json_record)
+
+    try:
+        with db.session.begin_nested():
+            record = InspireRecord.create_or_update(json_record)
+            record.commit()
+    except ValidationError as exc:
+        pattern = u"Migrator Validator Error: {}, Value: %r, Record: %r"
+        LOGGER.error(
+            pattern.format(".".join(exc.schema_path)), exc.instance, prod_record.recid
```
Comment (pazembrz) [inline inspirehep/migrator/tasks.py]: Use f-strings not .format() method

File: inspirehep/migrator/utils.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+def ensure_valid_schema(record):
+    """Make sure the ``$schema`` key of the record is valid.
+
+    This is done by setting the correct url to the schema, in case it only
+    contains the schema filename.
+    """
+    if not record["$schema"].startswith("http"):
+        record["$schema"] = url_for(
+            "invenio_jsonschemas.get_schema",
+            schema_path="records/{0}".format(record["$schema"]),
```
Comment (pazembrz) [inline inspirehep/migrator/utils.py]: Use f-strings not .format() method

File: inspirehep/migrator/models.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+from zlib import compress, decompress, error
+
+from invenio_db import db
+from sqlalchemy.ext.hybrid import hybrid_property
+
+from .utils import get_collection_from_marcxml
+
+
+class LegacyRecordsMirror(db.Model):
+    __tablename__ = "legacy_records_mirror"
```
Comment (pazembrz) [inline inspirehep/migrator/models.py]: Remove `legacy_records_mirror` from ALEMBIC_SKIP_TABLES in config, as you are adding this table to inspirehep, so we want to start managing it in our alembic.

Comment (pazembrz) [inline inspirehep/migrator/models.py]: Remove legacy_records_mirror from ALEMBIC_SKIP_TABLES in config, as you are adding this table to inspirehep, so we want to start managing it in our alembic.

File: inspirehep/migrator/utils.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    collections = set()
+    for field in force_list(marc_record.get("980__")):
+        for v in field.values():
+            for e in force_list(v):
+                collections.add(e.upper().strip())
+    if "DELETED" in collections:
+        return "DELETED"
+    for collection in collections:
+        if collection in REAL_COLLECTIONS:
+            return collection
```
Comment (pazembrz) [inline inspirehep/migrator/utils.py]: Maybe instead last for loop do:
```suggestion
            collections &= set(REAL_COLLECTIONS)
            return collections.pop() if collections else "HEP"
```
WDYT?

File: inspirehep/migrator/tasks.py [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+        finally:
+            lock.release()
+    else:
+        LOGGER.info("Continuous_migration already executed. Skipping.")
+
+
+@shared_task(ignore_result=False, queue="migrator")
+@disable_orcid_push
+def migrate_recids_from_mirror(prod_recids, skip_files=False):
+    models_committed.disconnect(index_after_commit)
```
Comment (michamos) [inline inspirehep/migrator/tasks.py]: is this still how indexing in done?

File: inspirehep/records/api/base.py [hunk 10]
```diff
@@ -420,7 +420,12 @@ def redirect(self, other):
     def create_or_update(cls, data, **kwargs):
         control_number = data.get("control_number")
         try:
-            record = cls.get_record_by_pid_value(control_number)
+            record_class = cls.get_class_for_record(data)
```
[the remark this answers is not in the document]
Comment (drjova) [reply inspirehep/records/api/base.py]: ``get_record_by_pid_value`` doesn't have the record ``data``

**Comment by drjova** [comment]
@michamos @pazembrz updated

File: docker-compose.core.yml [hunk 11]
```diff
@@ -61,7 +61,7 @@ services:
     extends:
       file: docker-compose.services.yml
       service: app
-    command: poetry run watchmedo auto-restart -d . -p '*.py' -- celery worker -E -B -A inspirehep.celery -l INFO
+    command: poetry run watchmedo auto-restart -d . -p '*.py' -- celery worker -E -A inspirehep.celery -l INFO --purge --queues celery,migrator,indexer_task
```
Comment (michamos) [inline docker-compose.core.yml]: names of the queues are the same as for legacy. Does anything need to be done to be sure the inspire-next workers don't comsume these queues any more?

File: inspirehep/migrator/cli.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    populate_mirror_from_file,
+)
+
+
+def halt_if_debug_mode(force):
+    message = """\
+    The application is running in debug mode, which leaks memory when doing
+    many database operations. To avoid problems, disable debug mode. This can
+    be done by setting "DEBUG=False" in the config or setting the environment
+    variable "APP_DEBUG=False". If you know what you are doing, you can pass
```
Comment (michamos) [inline inspirehep/migrator/cli.py]: is the `APP_` prefix still in use for env variables or is it `INVENIO_` now?

File: inspirehep/migrator/cli.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+@migrate.command("file")
+@click.argument(
+    "file_name", type=click.Path(exists=True, dir_okay=False, resolve_path=True)
+)
+@click.option(
+    "--mirror-only",
+    "-m",
+    is_flag=True,
+    default=False,
+    help="Only mirror the records without instead of doing a full migration.",
```
Comment (michamos) [inline inspirehep/migrator/cli.py]: ~~`without`~~

File: inspirehep/migrator/cli.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+)
+@click.option(
+    "--wait",
+    "-w",
+    is_flag=True,
+    default=False,
+    help="Wait for migration to complete. This only has an effect if the -w flag is not set.",
```
Comment (michamos) [inline inspirehep/migrator/cli.py]: `-m` instead of `-w`

File: inspirehep/migrator/cli.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    "also_migrate",
+    flag_value="broken",
+    help="Also migrate broken records, which did not migrate correctly in the previous run.",
+)
+@click.option(
+    "--wait",
+    "-w",
+    is_flag=True,
+    default=False,
+    help="Wait for migration to complete. This only has an effect if the -w flag is not set.",
```
Comment (michamos) [inline inspirehep/migrator/cli.py]: Here you can remove the second sentence.

File: inspirehep/migrator/cli.py (hunk 12 above)
Comment (drjova) [reply inspirehep/migrator/cli.py]: both of them are working

File: docker-compose.core.yml (hunk 11 above)
Comment (drjova) [reply docker-compose.core.yml]: we'll remove the queues from ``inspire-next`` workers when we switch migrator and indexer in inspirehep.

**Comment by drjova** [comment]
@michamos @pazembrz updated

**Pull request merged**